lib/web_ui/lib/src/engine/compositor/text.dart

```
@@ -0,0 +1,488 @@
+        html.window.console.warn(
+            'We do not support $boxHeightStyle. Defaulting to BoxHeightStyle.tight');
+        heightStyle = canvasKit['RectHeightStyle']['Tight'];
+        break;
+    }
+
+    js.JsObject widthStyle;
+    switch (boxWidthStyle) {
+      case ui.BoxWidthStyle.tight:
+        widthStyle = canvasKit['RectWidthStyle']['Tight'];
+        break;
+      case ui.BoxWidthStyle.max:
+        widthStyle = canvasKit['RectWidthStyle']['Max'];
+        break;
+    }
+
+    List<js.JsObject> skRects =
+        skParagraph.callMethod('getRectsForRange', <dynamic>[
+      start,
+      end,
+      heightStyle,
+      widthStyle,
+    ]);
+
+    List<ui.TextBox> result = List<ui.TextBox>(skRects.length);
+
+    for (int i = 0; i < skRects.length; i++) {
+      final js.JsObject rect = skRects[i];
+      result[i] = ui.TextBox.fromLTRBD(
+        rect['fLeft'],
+        rect['fTop'],
+        rect['fRight'],
+        rect['fBottom'],
+        _textDirection,
+      );
+    }
+
+    return result;
+  }
+
+  @override
+  ui.TextPosition getPositionForOffset(ui.Offset offset) {
+    js.JsObject positionWithAffinity =
+        skParagraph.callMethod('getGlyphPositionAtCoordinate', <double>[
+      offset.dx,
+      offset.dy,
+    ]);
+    return fromPositionWithAffinity(positionWithAffinity);
+  }
+
+  @override
+  ui.TextRange getWordBoundary(ui.TextPosition position) {
+    js.JsObject skRange =
+        skParagraph.callMethod('getWordBoundary', <int>[position.offset]);
+    return ui.TextRange(start: skRange['start'], end: skRange['end']);
+  }
+
+  @override
+  void layout(ui.ParagraphConstraints constraints) {
+    assert(constraints.width != null);
+    // TODO(het): CanvasKit throws an exception when laid out with
+    // a font that wasn't registered.
+    try {
+      skParagraph.callMethod('layout', <double>[constraints.width]);
+    } catch (e) {
+      html.window.console.warn('CanvasKit threw an exception while laying '
```

This should fail harder, no?

---

Changed to rethrow